File: go/279.go
package main

/*
 * This code can run but not work well
 * I still need some math analysis
 */
/* Recursive
func dp(n int, m map[int]int) int {
	if n == 0 {
		return 0
	} else if n == 1 {
		return 1
	} else if _, ok := m[n]; ok {
		return m[n]
	}
	m[n] = n
	for i := 2; i*i <= n; i++ {
		val := dp(n-i*i, m)
		if val < m[n] {
			m[n] = val
		}
	}
	m[n]++
	return m[n]
}

func numSquares(n int) int {
	return dp(n, make(map[int]int))
}
*/

/* Iterative */
func numSquares(n int) int {
	dp := make([]int, n+1)
	for i := 0; i <= n; i++ {
		dp[i] = i
	}

	for i := 1; i <= n; i++ {
		for j := 1; j*j <= i; j++ {
			if dp[i-j*j]+1 < dp[i] {
				dp[i] = dp[i-j*j] + 1
			}
		}
	}

	return dp[n]
}

File: cpp/279.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

/*
 * This code can run, but performance is bad,
 * I think I still need some math analysis skill
 */
class SolutionOld {
public:
  int dp(int n, unordered_map<int, int> &mp) {
    if (n == 0)
      return 0;
    if (mp.find(n) != mp.end())
      return mp[n];

    mp[n] = n;
    for (int i = 1; i * i <= n; i++) {
      int val = dp(n - i * i, mp);
      mp[n] = (mp[n] < val) ? mp[n] : val;
    }
    ++mp[n];
    return mp[n];
  }

  int numSquares(int n) {
    unordered_map<int, int> mp;
    dp(n, mp);
    return mp[n];
  }
};

// interative method
class Solution {
public:
  int numSquares(int n) {
    vector<int> arr(n + 1);
    for (int i = 0; i < arr.size(); i++)
      arr[i] = i;

    for (int i = 0; i <= n; i++)
      for (int j = 1; j * j <= i; j++)
        if (arr[i - j * j] + 1 < arr[i])
          arr[i] = arr[i - j * j] + 1;

    return arr[n];
  }
};
